### ISSUE TITLE
 ​used-car-price-prediction

### ISSUE THREAD

**Issue by muattarnaz7-svg** [muattarnaz7-svg/-used-car-price-prediction#1]
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import joblib

from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV, KFold
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error

sns.set_theme(style="whitegrid")
RANDOM_STATE = 42

def generate_car_data(n=1500, seed=RANDOM_STATE):
    """
    Simulates a realistic used-car listing dataset, loosely modeled on
    price behavior seen in local markets (PakWheels-style):
      - Newer cars cost more, but price drops fast in the first few years
      - Higher mileage lowers price
      - Certain brands/models hold value better than others
      - Automatic transmission and imported cars command a premium
    """
    rng = np.random.default_rng(seed)

    brands_models = {
        "Toyota Corolla": 4200000,
        "Honda Civic": 4800000,
        "Suzuki Alto": 1900000,
        "Suzuki Cultus": 2400000,
        "Honda City": 3600000,
        "Toyota Yaris": 3800000,
        "KIA Sportage": 6500000,
        "Hyundai Tucson": 6800000,
    }

    rows = []
    current_year = 2026

    for _ in range(n):
        model = rng.choice(list(brands_models.keys()))
        base_price = brands_models[model]

        year = rng.integers(2012, current_year)
        age = current_year - year

        mileage_km = max(2000, int(rng.normal(age * 12000, 8000)))

        engine_cc = rng.choice([1000, 1300, 1500, 1800, 2000],
                               p=[0.25, 0.25, 0.25, 0.15, 0.10])

        transmission = rng.choice(["Manual", "Automatic"], p=[0.55, 0.45])
        fuel_type = rng.choice(["Petrol", "Hybrid"], p=[0.9, 0.1])
        city = rng.choice(["Karachi", "Lahore", "Islamabad", "Faisalabad"],
                          p=[0.35, 0.35, 0.2, 0.1])
        is_imported = rng.choice([0, 1], p=[0.85, 0.15])
        num_owners = rng.choice([1, 2, 3, 4], p=[0.45, 0.30, 0.15, 0.10])
        accident_history = rng.choice([0, 1], p=[0.85, 0.15])

        price = base_price
        price *= (0.90 ** age)                   
        price -= mileage_km * 15                      
        price += (engine_cc - 1000) * 250              
        price += 150000 if transmission == "Automatic" else 0
        price += 300000 if fuel_type == "Hybrid" else 0
        price += 400000 if is_imported == 1 else 0
        price -= (num_owners - 1) * 80000              
        price -= 250000 if accident_history == 1 else 0
        price += rng.normal(0, 120000)                 
        price = max(300000, price)                      

        rows.append([model, year, mileage_km, engine_cc, transmission,
                    fuel_type, city, is_imported, num_owners,
                    accident_history, round(price, -3)])

    df = pd.DataFrame(rows, columns=[
        "model", "year", "mileage_km", "engine_cc", "transmission",
        "fuel_type", "city", "is_imported", "num_owners",
        "accident_history", "price"
    ])
    return df

def run_eda(df):
    print("=" * 65)
    print("EXPLORATORY DATA ANALYSIS")
    print("=" * 65)
    print(f"Shape: {df.shape}")
    print(f"\nMissing values:\n{df.isnull().sum()}")
    print(f"\nPrice summary:\n{df['price'].describe()}")

    fig, axes = plt.subplots(2, 2, figsize=(14, 10))

    sns.histplot(df["price"], kde=True, ax=axes[0, 0], color="#2563eb")
    axes[0, 0].set_title("Price Distribution")

    sns.scatterplot(data=df, x="mileage_km", y="price", hue="transmission",
                    alpha=0.6, ax=axes[0, 1])
    axes[0, 1].set_title("Mileage vs Price")

    sns.boxplot(data=df, x="model", y="price", ax=axes[1, 0])
    axes[1, 0].set_title("Price by Model")
    axes[1, 0].tick_params(axis="x", rotation=45)

    numeric_cols = df.select_dtypes(include="number")
    sns.heatmap(numeric_cols.corr(), annot=True, fmt=".2f",
               cmap="coolwarm", ax=axes[1, 1])
    axes[1, 1].set_title("Correlation Heatmap")

    plt.tight_layout()
    plt.savefig("eda_overview.png", dpi=120)
    plt.close()
    print("\nSaved: eda_overview.png")

def engineer_features(df):
    df = df.copy()
    current_year = 2026

    df["car_age"] = current_year - df["year"]
    df["km_per_year"] = df["mileage_km"] / df["car_age"].replace(0, 1)
    df["is_high_mileage"] = (df["mileage_km"] > df["mileage_km"].median()).astype(int)
    df["is_premium_brand"] = df["model"].isin(
        ["KIA Sportage", "Hyundai Tucson", "Honda Civic"]
    ).astype(int)

    return df

def build_preprocessor(numeric_features, categorical_features):
    numeric_transformer = Pipeline([("scaler", StandardScaler())])
    categorical_transformer = Pipeline(
        [("onehot", OneHotEncoder(handle_unknown="ignore"))]
    )

    preprocessor = ColumnTransformer([
        ("num", numeric_transformer, numeric_features),
        ("cat", categorical_transformer, categorical_features),
    ])
    return preprocessor

def compare_models(X_train, y_train, preprocessor):
    """
    Trains 3 different regression models and compares them using
    5-Fold Cross-Validation (not a single train/test split), scored
    by R^2. This gives a much more reliable performance estimate.
    """
    models = {
        "Linear Regression": LinearRegression(),
        "Random Forest": RandomForestRegressor(n_estimators=200, random_state=RANDOM_STATE),
        "Gradient Boosting": GradientBoostingRegressor(random_state=RANDOM_STATE),
    }

    kfold = KFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)
    results = {}

    print("\n" + "=" * 65)
    print("MODEL COMPARISON (5-Fold Cross-Validation, R^2 Score)")
    print("=" * 65)

    for name, model in models.items():
        pipe = Pipeline([("preprocessor", preprocessor), ("regressor", model)])
        scores = cross_val_score(pipe, X_train, y_train, cv=kfold, scoring="r2")
        results[name] = scores
        print(f"{name:<20}: mean R^2 = {scores.mean():.4f}  (+/- {scores.std():.4f})")

    plt.figure(figsize=(8, 5))
    plt.boxplot(results.values(), labels=results.keys())
    plt.title("Model Comparison - 5-Fold CV R^2 Scores")
    plt.ylabel("R^2 Score")
    plt.tight_layout()
    plt.savefig("model_comparison.png", dpi=120)
    plt.close()
    print("\nSaved: model_comparison.png")

    best_model_name = max(results, key=lambda k: results[k].mean())
    print(f"\nBest model based on CV: {best_model_name}")
    return best_model_name, models[best_model_name]

def tune_best_model(best_model_name, best_model, X_train, y_train, preprocessor):
    print("\n" + "=" * 65)
    print(f"HYPERPARAMETER TUNING: {best_model_name}")
    print("=" * 65)

    pipe = Pipeline([("preprocessor", preprocessor), ("regressor", best_model)])

    if best_model_name == "Random Forest":
        param_grid = {
            "regressor__n_estimators": [100, 200, 300],
            "regressor__max_depth": [None, 10, 20],
            "regressor__min_samples_split": [2, 5],
        }
    elif best_model_name == "Gradient Boosting":
        param_grid = {
            "regressor__n_estimators": [100, 200],
            "regressor__learning_rate": [0.05, 0.1],
            "regressor__max_depth": [3, 5],
        }
    else: 
        param_grid = {}

    if param_grid:
        grid = GridSearchCV(pipe, param_grid, cv=5, scoring="r2", n_jobs=-1)
        grid.fit(X_train, y_train)
        print(f"Best params: {grid.best_params_}")
        print(f"Best CV R^2: {grid.best_score_:.4f}")
        return grid.best_estimator_
    else:
        pipe.fit(X_train, y_train)
        return pipe

def evaluate_model(model, X_test, y_test):
    predictions = model.predict(X_test)

    mse = mean_squared_error(y_test, predictions)
    rmse = np.sqrt(mse)
    mae = mean_absolute_error(y_test, predictions)
    r2 = r2_score(y_test, predictions)

    print("\n" + "=" * 65)
    print("FINAL TEST SET EVALUATION")
    print("=" * 65)
    print(f"MAE  (avg error)      : Rs. {mae:,.0f}")
    print(f"RMSE                  : Rs. {rmse:,.0f}")
    print(f"R^2 Score             : {r2:.4f}")

    plt.figure(figsize=(7, 6))
    plt.scatter(y_test, predictions, alpha=0.5, color="#0ea5e9")
    lims = [min(y_test.min(), predictions.min()), max(y_test.max(), predictions.max())]
    plt.plot(lims, lims, "r--", label="Perfect Prediction")
    plt.xlabel("Actual Price (Rs.)")
    plt.ylabel("Predicted Price (Rs.)")
    plt.title("Actual vs Predicted Car Prices")
    plt.legend()
    plt.tight_layout()
    plt.savefig("actual_vs_predicted.png", dpi=120)
    plt.close()
    print("Saved: actual_vs_predicted.png")

    return {"mae": mae, "rmse": rmse, "r2": r2}

def plot_feature_importance(model, numeric_features, categorical_features):
    regressor = model.named_steps["regressor"]

    if not hasattr(regressor, "feature_importances_"):
        print("\nSelected model does not expose feature importances "
              "(e.g. Linear Regression) - skipping this step.")
        return

    ohe = model.named_steps["preprocessor"].named_transformers_["cat"].named_steps["onehot"]
    cat_feature_names = list(ohe.get_feature_names_out(categorical_features))
    all_feature_names = numeric_features + cat_feature_names

    importances = regressor.feature_importances_
    importance_df = pd.DataFrame({
        "feature": all_feature_names,
        "importance": importances
    }).sort_values("importance", ascending=False).head(12)

    plt.figure(figsize=(9, 6))
    sns.barplot(data=importance_df, x="importance", y="feature", palette="viridis",
               hue="feature", legend=False)
    plt.title("Top Feature Importances")
    plt.xlabel("Importance")
    plt.tight_layout()
    plt.savefig("feature_importance.png", dpi=120)
    plt.close()
    print("Saved: feature_importance.png")

    print("\nTop 5 most important features:")
    print(importance_df.head(5).to_string(index=False))

if __name__ == "__main__":
    df = generate_car_data(n=1500)
    run_eda(df)

    df = engineer_features(df)

    numeric_features = ["car_age", "mileage_km", "engine_cc", "km_per_year",
                        "num_owners", "is_high_mileage", "is_premium_brand",
                        "is_imported", "accident_history"]
    categorical_features = ["model", "transmission", "fuel_type", "city"]

    X = df[numeric_features + categorical_features]
    y = df["price"]

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=RANDOM_STATE
    )

    preprocessor = build_preprocessor(numeric_features, categorical_features)

    best_model_name, best_model = compare_models(X_train, y_train, preprocessor)

    final_model = tune_best_model(best_model_name, best_model, X_train, y_train, preprocessor)

    metrics = evaluate_model(final_model, X_test, y_test)

    plot_feature_importance(final_model, numeric_features, categorical_features)

    joblib.dump(final_model, "car_price_predictor.pkl")
    print("\n" + "=" * 65)
    print(f"Model saved successfully as 'car_price_predictor.pkl'")
    print(f"Final model: {best_model_name} | Test R^2: {metrics['r2']:.4f}")
    print("=" * 65)
